### PR TITLE
sinclair/chloe.cpp: Chloe 280SE (Timex TS2068 successor)- New WORKING

### PR DESCRIPTION
**Pull request by holub**
Credits: Andrew Owen

Chloe 280SE
Some aspect can be improved in the future.
ULA is generalization based on implementation in The Next branch - when any of it merged, I'll update the other one.

### CODE REVIEW

File: src/mame/sinclair/chloe.cpp [hunk 1]
```diff
@@ -0,0 +1,542 @@
...
+	Chloe 280SE
+**********************************************************************/
+
+#include "emu.h"
+
+#include "screen_ula.h"
+#include "spec128.h"
+
+#include "machine/spi_sdcard.h"
+#include "screen.h"
```
Comment (rb6502) [inline src/mame/sinclair/chloe.cpp:13]: Group includes with a directory prefix together, so machine/ and sound/ then screen and speaker.  Other than that this looks clean.

Comment (holub) [reply src/mame/sinclair/chloe.cpp:13]: Let me finish something more than that and we can potentially change it to WORKING

**Comment by holub** [comment]
Sorry for going back and forth. Let's draft it again as I about to add more things to it.

**Comment by holub** [comment]
Let's do it!

File: src/mame/sinclair/chloe.cpp [hunk 2]
```diff
@@ -118,33 +119,54 @@ void chloe_state::update_memory()
...
-		if (i == 0 && (m_divmmc_paged || !paged))
+		if (i == 0 && (divmmc_rom_active || !paged))
 		{
-			m_bank0_view.select(0);
-			if (m_divmmc_paged)
-				pg = 12;
+			if (divmmc_rom_active)
+			{
+				if (!mapram_mode || conmem)
+					pg = 24;
```
Comment (rb6502) [inline src/mame/sinclair/chloe.cpp:164]: The rest of this file looks to be in Allman brace style, so this if should also have braces.

File: src/mame/sinclair/chloe.cpp [hunk 3]
```diff
@@ -118,33 +119,54 @@ void chloe_state::update_memory()
...
 		{
-			if (m_divmmc_paged)
+			if (divmmc_rom_active)
 			{
-				pg = 16;
-				m_bank_ram[1]->set_entry((pg << 1) + (m_divmmc_reg & 0x0f));
+				pg = 32 + (divmmc_sram_page & 0x0f);
+				if (!mapram_mode || conmem)
+				{
+					if (!divmmc_sram_page_is_valid)
```
Comment (rb6502) [inline src/mame/sinclair/chloe.cpp:182]: Here again, needs braces.

File: src/mame/sinclair/chloe.cpp [hunk 4]
```diff
@@ -118,33 +119,54 @@ void chloe_state::update_memory()
...
+						m_bank1_view.select(0);
+				}
+				else
+				{
+					if ((mapram_mode && (divmmc_sram_page == 3)) || !divmmc_sram_page_is_valid)
```
Comment (rb6502) [inline src/mame/sinclair/chloe.cpp:187]: And here with the braces.

File: src/mame/sinclair/chloe.cpp [hunk 5]
```diff
@@ -193,7 +215,8 @@ void chloe_state::port_1ffd_w(u8 data)
 	if (m_port_7ffd_data & 0x20)
```
Comment (rb6502) [inline src/mame/sinclair/chloe.cpp]: Also here for braces.

File: src/mame/sinclair/chloe.cpp [hunk 6]
```diff
@@ -206,7 +229,16 @@ void chloe_state::port_ff_w(u8 data)
...
 void chloe_state::port_f4_w(u8 data)
 {
-	m_port_f4_data = data;
+	m_timex_mmu = data;
+	update_memory();
+}
+
+void chloe_state::port_e3_w(u8 data)
+{
+	if (m_divmmc_ctrl & 0x40)
```
Comment (rb6502) [inline src/mame/sinclair/chloe.cpp:256]: Here too for braces.

File: src/mame/sinclair/chloe.cpp [hunk 7]
```diff
@@ -242,6 +243,14 @@ void chloe_state::port_e3_w(u8 data)
 	update_memory();
 }
 
+void chloe_state::ay_address_w(u8 data)
+{
+	if ((data & 0xfe) == 0xfe)
```
Comment (rb6502) [inline src/mame/sinclair/chloe.cpp:265]: As mentioned in other comments, add braces here for consistency.

File: src/mame/sinclair/chloe.cpp [hunk 8]
```diff
@@ -448,12 +457,199 @@ void chloe_state::map_regs(address_map &map)
...
+{
+	u16 data = 0xffff;
+	u16 shifts = 0xffff;
+
+	u8 oi = offset >> 8;
+	for (u8 i = 0; i < 8; i++, oi >>= 1)
+	{
+		u16 line_data = m_io_line[i]->read();
+		shifts &= line_data;
+		if ((oi & 1) == 0)
```
Comment (rb6502) [inline src/mame/sinclair/chloe.cpp:580]: More missing braces.

File: src/mame/sinclair/chloe.cpp [hunk 9]
```diff
@@ -448,12 +457,199 @@ void chloe_state::map_regs(address_map &map)
...
+	}
+
+	bool shift_hold = ~m_io_line[0]->read() & 0x8000;
+	if (shift_hold && ((shifts & 0x1f00) != 0x1f00))
+	{
+		data >>= 8;
+		shifts >>= 8;
+	}
+
+	if (((offset & 0x0100) == 0) && BIT(~shifts, 6))
```
Comment (rb6502) [inline src/mame/sinclair/chloe.cpp:591]: Missing braces here and the next one.

File: src/mame/sinclair/chloe.cpp [hunk 10]
```diff
@@ -448,12 +457,199 @@ void chloe_state::map_regs(address_map &map)
...
+		data &= ~0x01; // CS
+
+	if (((offset & 0x8000) == 0) && BIT(~shifts, 7))
+		data &= ~0x02; // SS
+
+	data |= 0xe0;
+	data ^= 0x40;
+
+	/* cassette input from wav */
+	if (m_cassette->input() > 0.0038 )
```
Comment (rb6502) [inline src/mame/sinclair/chloe.cpp:599]: Missing braces.

File: src/mame/sinclair/chloe.cpp [hunk 11]
```diff
@@ -417,6 +419,12 @@ void chloe_state::raster_irq_adjust()
 		m_irq_raster_on_timer->reset();
 }
 
+INPUT_CHANGED_MEMBER(chloe_state::on_divmmc_nmi)
+{
+	if ((newval & 1) && (~m_io_line[0]->read() & 0x8000))
```
Comment (rb6502) [inline src/mame/sinclair/chloe.cpp:424]: Missing braces.

File: src/mame/sinclair/chloe.cpp [hunk 12]
```diff
@@ -346,13 +361,91 @@ u8 chloe_state::divmmc_disable_r(offs_t offset)
...
+}
+
+void chloe_state::raster_irq_adjust()
+{
+	if (BIT(m_uno_regs_data[0x0d], 1)) {
+		u16 line = (BIT(m_uno_regs_data[0x0d], 0) << 8) | m_uno_regs_data[0x0c];
+		m_irq_raster_on_timer->adjust(m_screen->time_until_pos((SCR_256x192.top() + line) % m_screen->height()));
+	}
+	else
+		m_irq_raster_on_timer->reset();
```
Comment (rb6502) [inline src/mame/sinclair/chloe.cpp:419]: This else clause should have braces.

**Comment by rb6502** [comment]
Ok.  A little cleanup and I think you're good.  I'd do it myself and save you the trouble, but editing PRs on other people's branches is a little beyond my Github skills at the moment.

**Comment by holub** [comment]
@rb6502 shall we?

**Comment by rb6502** [comment]
Sorry.  Let's do this :-)

**Pull request merged**